internal/charmstore/store.go

```diff
@@ -561,14 +561,17 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
 	return nil
 }
 
-// denormalizeEntity sets all denormalized fields in e
+// DenormalizeEntity sets all denormalized fields in e
 // from their associated canonical fields.
 //
 // It is the responsibility of the caller to set e.SupportedSeries
 // if the entity URL does not contain a series. If the entity
 // URL *does* contain a series, e.SupportedSeries will
 // be overwritten.
-func denormalizeEntity(e *mongodoc.Entity) {
+//
+// This is exported for the purposes of tests that
+// need to create directly into the database.
+func DenormalizeEntity(e *mongodoc.Entity) {
```

Comment:
As long as there are tests outside of internal/charmstore that rely on writing entities to the database directly, I think it's reasonable to do so. We're not exposing it to the outside world after all.